### DISCUSSION TITLE
Why session doesn't update or change when inside a loop?

### DISCUSSION THREAD

**Discussion by 4n1qz5skwv** [miguelgrinberg/Flask-SocketIO#1638] [Q&A] [answered]
I have multiple event handlers, Session value get changed in a handler, other handlers can access the updated value and also able to change it successfully, Which can be successfully accessible from other handlers.

Problem happen, when i wanted to stop a "for" loop, running inside a handler. So i made an another handler which set a stop value in session. Then i tried to access it inside that "for" loop. But the loop can't see the updated value. 

Then i tried to check if session was actually updated successfully by accessing it from other handlers. And yes it changed, i got the updated value. Only "for" loop can't see this. 

I tried to change value inside the "for" loop, it also didn't have any effect on session. other handler can't see that change. but only "for" loop can see its changed value.

It's like "for" loop using a virtual session of its own. copy the main session before entering the loop, then disconnect it self completely with the session.

Notes, I'm using filesystem session, session.modified is set after every change, all library is up to date. manage_session is False...

If my words are not clear enough, I'll make a mwe.

**Comment by miguelgrinberg** [comment, marked as answer]
There is nothing special about a for loop. The only thing that can prevent the session from being saved is an exception, which would interrupt your handler.

**Reply by 4n1qz5skwv** [reply]
```python
from flask import Flask, session, request
from flask_socketio import Namespace
from flask_socketio import SocketIO

from flask_session import Session

app = Flask(__name__)

app.config['SECRET_KEY'] = 'secret!'
app.config['SESSION_TYPE'] = 'filesystem'

sio = SocketIO(app, manage_session=False)
_sess = Session()
_sess.init_app(app)


@app.get('/')
def index():
    session['foobar'] = 'A'
    return """
<!DOCTYPE html>
<html>
<head>
<title>Page Title</title>
</head>
<body>

<h1>This is a Heading</h1>
<p>This is a paragraph.</p>
<button type="button" id="stop" >Stop</button> 
<button type="button" id="check" >Check</button> 
<script src="https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.min.js" integrity="sha256-/xUj+3OJU5yExlq6GSYGSHk7tPXikynS7ogEvDej/m4=" crossorigin="anonymous"></script>
<script src="https://cdn.jsdelivr.net/npm/socket.io-client@4.1.2/dist/socket.io.min.js" integrity="sha256-crRzk/tlAssAtFB2iE3/ieE1DgXDrc9lW+3rvBM1Wqc=" crossorigin="anonymous"></script>
<script type="text/javascript" charset="utf-8">

var sio = io('/index', {
    autoConnect: true,
    reconnection: false,
    auth: {
        token: 'foobar'
    },
    transports: ['websocket']
});

sio.on("connect", () => {
    sio.emit('start');
});

$('#stop').on( "click", function() {
    console.log('Stop Clicked?')
    sio.emit('stop');
});
$('#check').on( "click", function() {
    console.log('Check Clicked?')
    sio.emit('check');
});

</script>

</body>
</html>

    """


# noinspection PyMethodMayBeStatic
class IndexSio(Namespace):
    def on_connect(self, auth: dict):
        print(session[auth['token']])

        session[request.sid] = session.pop(auth['token'], None)
        print(f'on_connect -> get {session[request.sid]}')
        if session[request.sid] is not None:
            session[request.sid] = 'on_connect'
            print(f'on_connect -> set {session[request.sid]}')
            return True
        else:
            return False

    def on_start(self):
        for i in range(100):
            sio.sleep(3)
            print(f'on_start -> get {session[request.sid]}')
            session[request.sid] = 'on_start'
            print(f'on_start -> set {session[request.sid]}')

    def on_stop(self):
        print(f'on_stop -> get {session[request.sid]}')
        session[request.sid] = 'on_stop'
        print(f'on_stop -> set {session[request.sid]}')

    def on_check(self):
        print(f'on_check -> get {session[request.sid]}')


sio.on_namespace(IndexSio('/index'))

if __name__ == '__main__':
    sio.run(app)

```

### Print:

```
A
on_connect -> get A
on_connect -> set on_connect
on_start -> get on_connect
on_start -> set on_start
on_stop -> get on_connect
on_stop -> set on_stop
on_start -> get on_start
on_start -> set on_start
on_stop -> get on_stop
on_stop -> set on_stop
on_start -> get on_start
on_start -> set on_start
on_start -> get on_start
on_start -> set on_start
on_check -> get on_stop
on_check -> get on_stop
on_check -> get on_stop
on_start -> get on_start
on_start -> set on_start
on_check -> get on_stop
on_stop -> get on_stop
on_stop -> set on_stop
on_start -> get on_start
on_start -> set on_start
on_check -> get on_stop
on_check -> get on_stop
on_stop -> get on_stop
on_stop -> set on_stop
on_start -> get on_start
on_start -> set on_start
on_start -> get on_start
on_start -> set on_start

Process finished with exit code -1

```

@miguelgrinberg

**Reply by miguelgrinberg** [reply]
The session is saved when the handler ends. As I said above, when the loop ends the handler will reach the end, and at that point the session will be saved.

**Reply by 4n1qz5skwv** [reply]
Oh. I See. **Thanks** for explaining. I thought, because session is using filesystem, writing and reading will happen directly to and from the disk. Thread, multi worker... shouldn't matter. [Flask-Caching](https://flask-caching.readthedocs.io/en/latest/api.html#flask_caching.Cache.set) work this way. 

Anyway, what method would you recommend to stop that loop? Or is it a bad practice to use the loop like that?

**Reply by miguelgrinberg** [reply]
I don't have the context to answer. A loop in itself is not a problem, but you are using the user session in a strange way. Normally a user session allows an event to store some data that can be retrieved in the future by another event. You are expecting a real-time update of the session to use it as a way of communicating between two running events in parallel. You may want to try using a global variable for that.
